### PR TITLE
Fix problems of tests for Python version 3.7, 3.8 and 3.9

### LINKED ISSUE #316: Support for Python 3.7 and above
# 🚀 Feature Request

Current CI tests would fail for Python 3.7 or above. We should support more recent versions of Python.

## Motivation

Python 3.7 is the earliest version of Python still getting fixes of any kind.

## Pitch

Python versions greater than 3.7 should be added to the CI:

https://github.com/pykale/pykale/blob/3724843b3a3abad83be8b32bde0ca4bacfbddea8/.github/workflows/test.yml#L22

The resulting failures should be debugged and dependencies updated etc. as needed.

### PR DESCRIPTION
**Pull request by sz144**
Fixes #316.

### Description
- Update to support 3.7 and above.
- Update PyTorch >= 1.11.0 to fix this [bug](https://github.com/Lightning-AI/lightning/issues/12324).
- Update PyG installation on colab in deepdta example to fix this [bug](https://github.com/pyg-team/pytorch_geometric/issues/3484).
- Increase notebook cell timeout limit from 300 to 3000 seconds to reduce timeout errors.

### Status
**Ready**

### Types of changes

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] In-line docstrings updated.
- [ ] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

File: setup.cfg
Comment (XianyuanLiu) [inline setup.cfg:9]: Is it necessary to update flake8 this time, in accordance with issue #323? This update mainly involves moving namespace from T0* to T2* to avoid collision with other libraries using the same error code.
If so, all maintainers need to reinstall flake by running `pip install -e .[dev]`.

Comment (sz144) [reply setup.cfg:9]: Shall we freeze the version 3.8.4 of flake8? It was released in Oct 2020, is it too old? I did not have any setup for pre-commit but printing errors will be reported here when upgrading to Python 3.7. We need to consider the compatibility for the future.

**Comment by sz144** [comment]
@haipinglu @XianyuanLiu @pz-white The errors reported in tests of Python 3.8 and 3.9  are caused by a problem between `pytorch_geometric` and `lightning`. There is a [PR](https://github.com/Lightning-AI/lightning/pull/14117) for `lightning` to fix it yesterday (9th Aug 2022) but not merged to the main branch yet. Shall we just disable the test on 3.8 and 3.9 and wait for the new release of `lightning`?

**Comment by pz-white** [comment]
Thanks, @sz144! I also found a similar conflict issue in lightning. I am trying whether we can remove PyG independence in the error test, since we actually do not need it in DeepDTA test. If not successful, we can consider to wait lightning fix. Btw, I found the lightning version is also old (1.7.1), will the new issue fix apply to our current used version?

**Comment by haipinglu** [comment]
> Thanks, @sz144! I also found a similar [conflict](https://github.com/Lightning-AI/lightning/issues/14050) issue in lightning five days ago. I am trying whether we can remove PyG independence in the error test, since we actually do not need it in DeepDTA test. If not successful, we can consider to wait lightning fix.

I prefer to make some minimal changes to avoid the error, e.g. commenting off a minimal number of tests, putting a note that this can be added back in future update (in test code as well as project board). Then we can move on to merge this with 3.8 and 3.9 green.

**Comment by haipinglu** [comment]
> @haipinglu @XianyuanLiu @pz-white The errors reported in tests of Python 3.8 and 3.9 are caused by a problem between `pytorch_geometric` and `lightning`. There is a [PR](https://github.com/Lightning-AI/lightning/pull/14117) for `lightning` to fix it yesterday (9th Aug 2022) but not merged to the main branch yet. Shall we just disable the test on 3.8 and 3.9 and wait for the new release of `lightning`?

Even merged, we may need to wait for the next release so that is not optimal. See my suggestion below to remove a minimal number of lines of tests.

**Comment by pz-white** [comment]
> > @haipinglu @XianyuanLiu @pz-white The errors reported in tests of Python 3.8 and 3.9 are caused by a problem between `pytorch_geometric` and `lightning`. There is a [PR](https://github.com/Lightning-AI/lightning/pull/14117) for `lightning` to fix it yesterday (9th Aug 2022) but not merged to the main branch yet. Shall we just disable the test on 3.8 and 3.9 and wait for the new release of `lightning`?
> 
> Even merged, we may need to wait for the next release so that is not optimal. See my suggestion below to remove a minimal number of lines of tests.

I see, I will remove relevant test lines and annotate to re-add in the future once the lightning issue is fixed. Thanks, Haiping!

**Comment by haipinglu** [comment]
> > @haipinglu @XianyuanLiu @pz-white The errors reported in tests of Python 3.8 and 3.9 are caused by a problem between `pytorch_geometric` and `lightning`. There is a [PR](https://github.com/Lightning-AI/lightning/pull/14117) for `lightning` to fix it yesterday (9th Aug 2022) but not merged to the main branch yet. Shall we just disable the test on 3.8 and 3.9 and wait for the new release of `lightning`?
> 
> Even merged, we may need to wait for the next release so that is not optimal. See my suggestion below to remove a minimal number of lines of tests. This should then become a general strategy to help us move forward more efficiently and survive in the ecosystem, since similar cases may still show up in future.



> > > @haipinglu @XianyuanLiu @pz-white The errors reported in tests of Python 3.8 and 3.9 are caused by a problem between `pytorch_geometric` and `lightning`. There is a [PR](https://github.com/Lightning-AI/lightning/pull/14117) for `lightning` to fix it yesterday (9th Aug 2022) but not merged to the main branch yet. Shall we just disable the test on 3.8 and 3.9 and wait for the new release of `lightning`?
> > 
> > 
> > Even merged, we may need to wait for the next release so that is not optimal. See my suggestion below to remove a minimal number of lines of tests.
> 
> I see, I will remove relevant test lines and annotate to re-add in the future once the lightning issue is fixed. Thanks, Haiping!

@pz-white Comment it/them off rather than remove please.

**Comment by otaj** [comment]
Hi, I'm a developer of lightning (and the one who broke your tests, sorry about that!), sadly this patch didn't make it into yesterday's patch release, however, it will be in master in about an hour (just waiting for tests to pass) and therefore it will be in next patch release (most likely next Tuesday, 2022-08-16)

**Comment by haipinglu** [comment]
> Hi, I'm a developer of lightning (and the one who broke your tests, sorry about that!), sadly this patch didn't make it into yesterday's patch release, however, it will be in master in about an hour (just waiting for tests to pass) and therefore it will be in next patch release (most likely next Tuesday, 2022-08-16)

Thank you @otaj for letting us know! That's really helpful! Lightning is fantastic and we learned and benefited a lot from it.

**Comment by otaj** [comment]
Btw, the lightning versions, which manifest the issue https://github.com/Lightning-AI/lightning/issues/14050 should only be `1.7.0` and `1.7.1`.

**Comment by haipinglu** [comment]
> Btw, the lightning versions, which manifest the issue [Lightning-AI/lightning#14050](https://github.com/Lightning-AI/lightning/issues/14050) should only be `1.7.0` and `1.7.1`.

That's good to know. Thanks a lot @otaj !

**Review by haipinglu — APPROVED** [review]
LGTM. Ready for merge after @pz-white double-check deepdta example changes.

**Review by pz-white — APPROVED** [review]
LGTM of the dti changes. Thanks, Haiping!

**Pull request merged**
